### PR TITLE
Prepare translate context for server

### PR DESCRIPTION
**Pull request by Pengpengwanga**
Signed-off-by: Pengpeng Wang <pengpengw@vmware.com>

### CODE REVIEW

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 1]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+    * }
+    */
+   private Map<String, Map<String, String>> getToNameAndFromulasMapForServer(Asset server){
+      Map<String, Map<String, String>> toNameAndFormulasMap = new HashMap<String, Map<String, String>>();
+      Map<String, String> formulas = server.getMetricsformulars();
+      if(formulas == null || formulas.isEmpty()) {
+         return toNameAndFormulasMap;
+      }
+      //Server Host Usage Formula
+      Map<String, String> hostUsageformulas = server.getMetricsformulars();
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1298]: remove this line

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 2]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+      String hostUsageformulasInfo = formulas.get(FlowgateConstant.HOST_METRICS);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1299]: better extract the getxformular logic.

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 3]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+      if (!StringUtils.isBlank(hostUsageformulasInfo)) {
+         hostUsageformulas =
+               server.metricsFormulaToMap(hostUsageformulasInfo, new TypeReference<Map<String, String>>(){});
+         for(Map.Entry<String, String> metricNameAndFormula : hostUsageformulas.entrySet()) {
+            Map<String, String> metricNameAndFormulaMap = new HashMap<>();
+            String toName = metricNameAndFormula.getKey();
+            metricNameAndFormulaMap.put(metricNameAndFormula.getKey(), metricNameAndFormula.getValue());
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1306]: we should first check the translation table

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 4]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+         pduPorts = Arrays.asList(allPduPortInfo.split(FlowgateConstant.SPILIT_FLAG));
+         pduAssetIdAndUsedOutletMap = new HashMap<String, String>();
+         for (String pduPortInfo : pduPorts) {
+            // startport_FIELDSPLIT_endDeviceName_FIELDSPLIT_endport_FIELDSPLIT_endDeviceAssetID
+            // item[0] start port
+            // item[1] device name
+            // item[2] end port
+            // itme[3] assetid
+            String items[] = pduPortInfo.split(FlowgateConstant.SEPARATOR);
+            pduAssetIdAndUsedOutletMap.put(items[3], items[2]);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1329]: consider server connect multiple outlet in same pdu.

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 5]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+      }
+      for(Map.Entry<String, Map<String, String>> pduIdAndFormula : pduFormulas.entrySet()) {
+         String pduId = pduIdAndFormula.getKey();
+         Map<String, String> formulaKeyAndFormulaMap = pduIdAndFormula.getValue();
+         String outLet = pduAssetIdAndUsedOutletMap.get(pduId);
+         for(Map.Entry<String, String> formulaEntry : formulaKeyAndFormulaMap.entrySet()) {
+            Map<String, String> metricNameAndFormulaMap = new HashMap<>();
+            String formulaKey = formulaEntry.getKey();
+            String formulString = formulaEntry.getValue();
+            String metricName = formulaKeyAndMetricNameMap.get(formulaKey);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1340]: we need to check if the formulakey not appeared in the translation mp

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 6]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+            // startport_FIELDSPLIT_endDeviceName_FIELDSPLIT_endport_FIELDSPLIT_endDeviceAssetID
+            // item[0] start port
+            // item[1] device name
+            // item[2] end port
+            // itme[3] assetid
+            String items[] = pduPortInfo.split(FlowgateConstant.SEPARATOR);
+            pduAssetIdAndUsedOutletMap.put(items[3], items[2]);
+         }
+      }
+      for(Map.Entry<String, Map<String, String>> pduIdAndFormula : pduFormulas.entrySet()) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1332]: add pduFormulas same here.

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 7]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+         String pduId = pduIdAndFormula.getKey();
+         Map<String, String> formulaKeyAndFormulaMap = pduIdAndFormula.getValue();
+         String outLet = pduAssetIdAndUsedOutletMap.get(pduId);
+         for(Map.Entry<String, String> formulaEntry : formulaKeyAndFormulaMap.entrySet()) {
+            Map<String, String> metricNameAndFormulaMap = new HashMap<>();
+            String formulaKey = formulaEntry.getKey();
+            String formulString = formulaEntry.getValue();
+            String metricName = formulaKeyAndMetricNameMap.get(formulaKey);
+            metricNameAndFormulaMap.put(metricName, formulString);
+            String toName = formulaKeyAndMetricFormatNameMap.get(formulaKey);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1342]: should check if get()==null .

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 8]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+    * {
+    *    toName:{metricName,formula}
+    * }
+    *
+    * Sample value : SERVER_CONNECTED_PDU_CURRENT
+    * {
+    *    PDU:pduID|Current:{Current,id1+id2}
+    * }
+    */
+   private Map<String, Map<String, String>> getToNameAndFromulasMapForServer(Asset server){
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1291]: getMetricDispalyNameAndFormularMapForServer

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 9]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+               List<ValueUnit> valueUnits = new ArrayList<>();
+               for(ValueUnit valueUnit : assetAndValueUnitsMap.get(id)) {
+                  if(!metricName.equals(valueUnit.getKey())) {
+                     continue;
+                  }
+                  String pdu_outlet_extraidentifier = valueUnit.getExtraidentifier();
+                  String outlet = null;
+                  switch (metricName) {
+                  case MetricName.PDU_CURRENT:
+                  case MetricName.PDU_APPARENT_POWER:
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1066]: may not the right place, but we should try to keep valueunit name and metric name as the same value. like pdu total current and current of the the outlet and inlet, they should be both called current, use extra-identifier can identify them .

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 10]
```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+               if(valueUnits.isEmpty()) {
+                  continue;
+               }
+               Collections.sort(valueUnits, comparator);
+               idAndValues.put(id, valueUnits);
+            }
+            //List<Map<String, ValueUnit>> valueUnitMaps: [[{id1, valueUnit1},{id2, valueUnit1}] , []]
+            //A idAndValueUnitsMap correspond a Metric-Data
+            //The list valueUnitMaps correspond the Metric-Data list
+            String IdAndValueUnitMapString = iDAndValueUnitMapListToString(idAndValues);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1108]: To be continue.

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 11]
```diff
@@ -956,4 +985,419 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+                  idAndValuesForServerVoltage.put(id, outletVoltages);
+               }
+               if(valueUnits.isEmpty()) {
+                  continue;
+               }
+               Collections.sort(valueUnits, comparator);
+               idAndValues.put(id, valueUnits);
+            }
+            if(!idAndValues.isEmpty()) {
+               metricDatas.addAll(translateToMetricDataForOneMetric(displayName, idFormula, idAndValues));
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1104]: formula

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 12]
```diff
@@ -956,4 +985,419 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+   private String getDisplayName(String formulaKey) {
+      String displayName = formulaKeyAndMetricFormatNameMap.get(formulaKey);
+      if(StringUtils.isEmpty(displayName)) {
+         displayName = formulaKey;
+      }
+      return displayName;
+   }
+
+   private String getMetricName(String formulaKey) {
+      String metricName = formulaKeyAndMetricNameMap.get(formulaKey);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1396]: if xxxmap.contain(key){
return xxmap.get(key)
}
return formulakey

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 13]
```diff
@@ -75,6 +76,11 @@
    //This map databaseUnitAndOutputUnitMap will be deprecated in Flowgate-1.3
    private static Map<String, String> databaseUnitAndOutputUnitMap = null;
    private static Map<String,String> metricNameMap = new HashMap<String,String>();
+   private static Map<String, String> formulaKeyAndMetricNameMap = new HashMap<>();
+   private static Map<String, String> formulaKeyAndMetricFormatNameMap = new HashMap<>();
+   private static String TONAME = "ToName";
```
[the remark this answers is not in the document]
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java (hunk 1 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java (hunk 4 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 14]
```diff
@@ -952,4 +978,435 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+      }
+      return metricDatas;
+   }
+
+   private List<MetricData> translateToMetricData(Map<String, List<ValueUnit>> assetAndValueUnitsMap,
+         Asset asset){
+      List<MetricData> metricDatas = new ArrayList<MetricData>();
+      Map<String, Map<String, String>> displayNameAndFormulasMap = getMetricDispalyNameAndFormularMapForServer(asset);
+      if(displayNameAndFormulasMap.isEmpty() ||
+            !displayNameAndFormulasMap.containsKey(DISPLAYNAMEANDFORMULANAMEMAPKEY)) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1010]: move the DISPLAYNAMEANDFORMULANAMEMAPKEY key before loop the map.

File: flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java [hunk 15]
```diff
@@ -952,4 +978,435 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
...
+                  valueUnits = outletVoltages;
+               }
+               if(valueUnits.isEmpty()) {
+                  continue;
+               }
+               Collections.sort(valueUnits, comparator);
+               idAndValues.put(id, valueUnits);
+            }
+            if(!idAndValues.isEmpty()) {
+               metricDatas.addAll(translateToMetricDataForOneMetric(displayName,displayNameAndFormulaKeyMap.get(displayName), formula, idAndValues));
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java:1101]: convertValueUnitToMetricData

**Review by yixingjia — APPROVED** [review]

**Pull request merged**